lib/AST/ASTMangler.cpp

```diff
@@ -1470,48 +1469,76 @@ void ASTMangler::appendFunctionType(AnyFunctionType *fn,
   }
 }
 
-void ASTMangler::appendFunctionSignature(AnyFunctionType *fn,
-                                         bool forceSingleParam) {
-  appendParams(fn->getResult(), /*forceSingleParam*/ false);
-  appendParams(fn->getInput(), forceSingleParam);
+void ASTMangler::appendFunctionSignature(AnyFunctionType *fn) {
+  appendFunctionResultType(fn->getResult());
+  appendFunctionInputType(fn->getParams());
   if (fn->throws())
     appendOperator("K");
 }
 
-void ASTMangler::appendParams(Type ParamsTy, bool forceSingleParam) {
-  if (TupleType *Tuple = ParamsTy->getAs<TupleType>()) {
-    if (Tuple->getNumElements() == 0) {
-      if (forceSingleParam) {
-        // A tuple containing a single empty tuple.
-        appendOperator("y");
-        appendOperator("t");
-        appendListSeparator();
-        appendOperator("t");
-      } else {
-        appendOperator("y");
-      }
-      return;
+void ASTMangler::appendFunctionInputType(
+    ArrayRef<AnyFunctionType::Param> params) {
+  auto getParamType = [](const AnyFunctionType::Param &param) -> Type {
+    auto type = param.getType();
+
+    // FIXME: Change mangling for variadic parameters so
+    //        the enclosing array type is not required.
+    if (param.isVariadic()) {
+      auto *arrayDecl = type->getASTContext().getArrayDecl();
+      assert(arrayDecl);
+      return BoundGenericType::get(arrayDecl, Type(), {type});
     }
-    if (forceSingleParam && Tuple->getNumElements() > 1) {
-      auto flags = ParameterTypeFlags();
-      if (ParenType *Paren = dyn_cast<ParenType>(ParamsTy.getPointer())) {
-        ParamsTy = Paren->getUnderlyingType();
-        flags = Paren->getParameterFlags();
-      }
 
-      appendType(ParamsTy);
-      if (flags.isShared())
+    return type;
+  };
+
+  switch (params.size()) {
+  case 0:
+    appendOperator("y");
+    break;
+
+  case 1: {
+    const auto &param = params.front();
+
+    // If this is just a single parenthesized type,
+    // to save space in the mangled name, let's encode
+    // it as a single type dropping sugar.
+    if (!param.hasLabel() && !param.isVariadic()) {
```

Comment:
Can't you just add a "&& !isa<TupleType>(param.getType().getPointer())" here and get rid of the if-statement below (line 5110)?
The tuple-case should be handled the same way in the default case

Reply:
That would be incorrect because checking label and the param type are two different things e.g. ((Int, Int)) is going to form a single parameter with (Int, Int) tuple inside, and checking for presence of label first makes sure that it wasn't a single element tuple with tuple inside e.g. (foo: (Int, Int)), and then if the type of the element itself is a tuple it adds enclosing container to the name to make sure that we don't drop the sugar, otherwise we just encode it as a regular single type.

Reply:
Thanks, @eeckstein ! I've made the change you suggested, didn't realize that there was duplication going on...